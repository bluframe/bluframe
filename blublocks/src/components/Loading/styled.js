// @flow
/* eslint-disable no-ternary */

import styled, { type StyledComponent, keyframes } from "styled-components"
import type { Theme } from "@bluframe/blublocks"
import getColorGrades from "helpers/getColorGrades"

type Props = {
  +$contrast?: boolean,
  +$secondary?: boolean,
  +disabled?: boolean,
  +size: number
}

const rotation = keyframes`
    0% {
        transform: rotate(0deg);
    }
    100% {
        transform: rotate(360deg);
    }
`

const getColor = (props: { ...Props, +theme: Theme }) => {
  const colorGrades = getColorGrades(props)

  return props.$contrast ? colorGrades.contrast : colorGrades.main
}

export const Wrapper: StyledComponent<
  Props,
  Theme,
  HTMLDivElement
> = styled.div`
  animation: ${rotation} 1s linear infinite;
  border: 2px solid ${getColor};
  border-bottom-color: transparent;
  border-radius: 50%;
  height: ${({ size }) => `${size}px`};
  width: ${({ size }) => `${size}px`};
`
