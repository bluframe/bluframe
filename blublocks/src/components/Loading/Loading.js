// @flow

import type { ComponentProps } from "."
import React from "react"
import { Wrapper } from "./styled"

const DEFAULT_SIZE = 16

const Loading = ({
  contrast,
  disabled,
  secondary,
  size = DEFAULT_SIZE
}: ComponentProps): React$Node => (
  <Wrapper
    contrast={contrast}
    disabled={disabled}
    secondary={secondary}
    size={size}
  />
)

export default Loading
