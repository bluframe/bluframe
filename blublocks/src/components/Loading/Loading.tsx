import { ComponentProps } from "."
import { Wrapper } from "./styled"

const DEFAULT_SIZE = 16

const Loading = ({
  contrast,
  disabled,
  secondary,
  size = DEFAULT_SIZE
}: ComponentProps) => (
  <Wrapper
    $contrast={contrast}
    $secondary={secondary}
    aria-label="Loading"
    disabled={disabled}
    role="status"
    size={size}
  />
)

export default Loading
