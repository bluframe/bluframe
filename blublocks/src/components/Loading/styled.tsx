/* eslint-disable no-ternary */

import styled, { keyframes } from "styled-components"
import { Theme } from "styles/theme"
import getColorGrades from "helpers/getColorGrades"

interface Props {
  $contrast?: boolean
  $secondary?: boolean
  disabled?: boolean
  size: number
}

const rotation = keyframes`
    0% {
        transform: rotate(0deg);
    }
    100% {
        transform: rotate(360deg);
    }
`

const getColor = (props: Props & { theme: Theme }) => {
  const colorGrades = getColorGrades(props)

  return props.$contrast ? colorGrades.contrast : colorGrades.main
}

export const Wrapper = styled.div<Props>`
  animation: ${rotation} 1s linear infinite;
  border: 2px solid ${getColor};
  border-bottom-color: transparent;
  border-radius: 50%;
  height: ${({ size }: { size: number }) => `${size}px`};
  width: ${({ size }: { size: number }) => `${size}px`};
`
