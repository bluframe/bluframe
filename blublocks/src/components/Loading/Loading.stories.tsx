import Loading, { Props } from "components/Loading"
import React from "react"

const stories = {
  argTypes: {
    contrast: { control: { type: "boolean" } },
    disabled: { control: { type: "boolean" } },
    secondary: { control: { type: "boolean" } },
    size: { control: { type: "number" } }
  },
  component: Loading,
  title: "Components/Loading"
}

const Template = (args: Props) => <Loading {...args} />

export const Default: any = Template.bind({})
Default.args = {}

export default stories
