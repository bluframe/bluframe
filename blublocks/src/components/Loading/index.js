// @flow

import Loading from "./Loading"
import type { LoadingProps } from "@bluframe/blublocks"

export type Props = {|
  ...LoadingProps
|}

export type ComponentProps = {|
  ...Props
|}

export default Loading
