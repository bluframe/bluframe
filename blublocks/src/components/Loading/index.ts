import Loading from "./Loading"

export interface Props {
  contrast?: boolean
  disabled?: boolean
  secondary?: boolean
  size?: number
}

export interface ComponentProps extends Props {}

export default Loading
