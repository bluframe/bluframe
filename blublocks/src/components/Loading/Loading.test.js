// @flow

import Loading from "./Loading"
import React from "react"
import { render } from "tests"

jest.mock(".", () => ({}))

describe("Loading", () => {
  it("renders with default props", () => {
    const { container } = render(<Loading />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders with custom size", () => {
    const { container } = render(<Loading size={32} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders with contrast color", () => {
    const { container } = render(<Loading contrast />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders with secondary color", () => {
    const { container } = render(<Loading secondary />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders in disabled state", () => {
    const { container } = render(<Loading disabled />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders with multiple custom props", () => {
    const { container } = render(
      <Loading contrast disabled secondary size={40} />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
