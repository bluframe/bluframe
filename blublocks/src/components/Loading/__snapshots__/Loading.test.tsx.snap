// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading renders in disabled state 1`] = `
<div
  aria-label="Loading"
  class="sc-bdDqHD gpOZpg"
  disabled=""
  role="status"
  size="16"
/>
`;

exports[`Loading renders with contrast color 1`] = `
<div
  aria-label="Loading"
  class="sc-bdDqHD fnoIOa"
  role="status"
  size="16"
/>
`;

exports[`Loading renders with custom size 1`] = `
<div
  aria-label="Loading"
  class="sc-bdDqHD bzCA-dM"
  role="status"
  size="32"
/>
`;

exports[`Loading renders with default props 1`] = `
<div
  aria-label="Loading"
  class="sc-bdDqHD kbywGc"
  role="status"
  size="16"
/>
`;

exports[`Loading renders with multiple custom props 1`] = `
<div
  aria-label="Loading"
  class="sc-bdDqHD kcwmPS"
  disabled=""
  role="status"
  size="40"
/>
`;

exports[`Loading renders with secondary color 1`] = `
<div
  aria-label="Loading"
  class="sc-bdDqHD kzYUtK"
  role="status"
  size="16"
/>
`;
